File: db/state.py
"""Vendor state machine (PRD Section 7).

Status changes happen ONLY through `transition()`, which validates the move
against the whitelist, enforces the human/LLM gates, appends an `events` row
and then mirrors the new status onto `vendors.status`.  `rebuild_status()`
replays the event log and asserts that the mirror is faithful.
"""
from __future__ import annotations

import json
import sqlite3
from datetime import datetime, timezone
from typing import Any

from db.init import get_conn

# --- vocabulary ---------------------------------------------------------------

STATUSES: tuple[str, ...] = (
    "Identified", "Screened", "Qualified", "Contacted", "Replied",
    "In Discussion", "Approved", "Rejected", "Dormant",
)
INITIAL_STATUS = "Identified"

# diligence_stage lives inside "In Discussion" and only ever moves forward
STAGES: tuple[str, ...] = ("responded", "technical_review", "quote_received", "sampling")
HUMAN_ONLY_STAGES = frozenset({"quote_received", "sampling"})
LLM_AUTO_STAGES = frozenset({"responded", "technical_review"})

# Transition whitelist (from -> allowed to).  "Any -> Approved / Rejected" is
# human-only; "Rejected / Dormant -> Qualified" is the human reopen path.
TRANSITIONS: dict[str, frozenset[str]] = {
    "Identified":    frozenset({"Screened", "Approved", "Rejected"}),
    "Screened":      frozenset({"Qualified", "Rejected", "Approved"}),
    "Qualified":     frozenset({"Contacted", "Rejected", "Approved"}),
    "Contacted":     frozenset({"Replied", "Dormant", "Rejected", "Approved"}),
    "Replied":       frozenset({"In Discussion", "Rejected", "Approved"}),
    "In Discussion": frozenset({"Approved", "Rejected"}),
    "Approved":      frozenset({"Rejected"}),
    "Rejected":      frozenset({"Qualified"}),
    "Dormant":       frozenset({"Qualified", "Approved", "Rejected"}),
}

# Gates.  Everything not listed here requires actor == "human".
AUTOMATIC: frozenset[tuple[str, str]] = frozenset({
    ("Identified", "Screened"),   # ruleset executed
    ("Contacted", "Replied"),     # inbound email, deterministic
    ("Contacted", "Dormant"),     # no reply in 10 days (P1)
})
LLM_ALLOWED: frozenset[tuple[str, str]] = frozenset({
    ("Replied", "In Discussion"),
    ("Replied", "Rejected"),
})
LLM_AUTO_THRESHOLD = 0.85
REOPEN: frozenset[tuple[str, str]] = frozenset({("Rejected", "Qualified"), ("Dormant", "Qualified")})


class IllegalTransition(ValueError):
    """Raised when a status/stage move is not in the whitelist."""


class GateViolation(PermissionError):
    """Raised when the actor is not allowed to perform an otherwise-legal move."""


class StatusMismatch(AssertionError):
    """Raised by rebuild_status when the event log disagrees with vendors.status."""


def utcnow() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="microseconds")


# --- helpers ------------------------------------------------------------------

def _is_legal(from_status: str | None, to_status: str, from_stage: str | None, to_stage: str | None) -> bool:
    if from_status is None:
        return to_status == INITIAL_STATUS and to_stage is None
    if from_status == to_status:
        # stage-only advance inside In Discussion
        if from_status != "In Discussion" or to_stage is None:
            return False
        if to_stage not in STAGES:
            return False
        cur = STAGES.index(from_stage) if from_stage in STAGES else -1
        return STAGES.index(to_stage) > cur
    if to_status not in TRANSITIONS.get(from_status, frozenset()):
        return False
    if to_stage is not None:
        # a stage may only be set when landing in In Discussion
        return to_status == "In Discussion" and to_stage in STAGES
    return True


def _check_gate(from_status: str, to_status: str, to_stage: str | None,
                actor: str, confidence: float | None, reason: str) -> None:
    if not reason or not str(reason).strip():
        raise GateViolation("a non-empty reason is required for every transition")
    if actor == "human":
        return
    pair = (from_status, to_status)
    if pair in REOPEN:
        raise GateViolation(f"{from_status} -> Qualified (reopen) is human-only")
    if to_stage in HUMAN_ONLY_STAGES:
        raise GateViolation(f"diligence_stage '{to_stage}' is always via the review queue (human)")
    confident = confidence is not None and float(confidence) >= LLM_AUTO_THRESHOLD
    if from_status == to_status:  # stage-only advance
        if actor == "llm_inference" and to_stage in LLM_AUTO_STAGES and confident:
            return
        raise GateViolation(
            f"stage advance to '{to_stage}' by '{actor}' requires human or llm_inference with confidence >= {LLM_AUTO_THRESHOLD}")
    if pair in AUTOMATIC:
        return
    if actor == "llm_inference" and pair in LLM_ALLOWED and confident:
        return
    raise GateViolation(
        f"{from_status} -> {to_status} by '{actor}' (confidence={confidence}) requires a human")


def current_state(vendor_id: str, conn: sqlite3.Connection | None = None) -> tuple[str, str | None]:
    """(status, diligence_stage) as recorded on vendors + the latest event."""
    conn = conn or get_conn()
    row = conn.execute("SELECT status FROM vendors WHERE vendor_id = ?", (vendor_id,)).fetchone()
    if row is None:
        raise KeyError(f"unknown vendor_id {vendor_id!r}")
    ev = conn.execute(
        "SELECT to_status, to_stage FROM events WHERE vendor_id = ? ORDER BY created_at DESC, event_id DESC LIMIT 1",
        (vendor_id,),
    ).fetchone()
    stage = ev["to_stage"] if ev is not None and ev["to_status"] == "In Discussion" else None
    return row["status"], stage


# --- public API ---------------------------------------------------------------

def register_vendor(vendor: dict[str, Any], actor: str, reason: str = "discovered",
                    conn: sqlite3.Connection | None = None, payload: dict | None = None) -> bool:
    """Insert a vendor in status Identified together with its creation event.

    Returns False (and changes nothing) if the vendor_id already exists.
    `vendor` holds the vendors columns except status/discovered_at/updated_at.
    """
    conn = conn or get_conn()
    if conn.execute("SELECT 1 FROM vendors WHERE vendor_id = ?", (vendor["vendor_id"],)).fetchone():
        return False
    now = utcnow()
    cols = {
        "vendor_id": vendor["vendor_id"],
        "name": vendor["name"],
        "vendor_type": vendor["vendor_type"],
        "primary_domain": vendor.get("primary_domain"),
        "country": vendor.get("country"),
        "contact_email": vendor.get("contact_email"),
        "attributes": json.dumps(vendor.get("attributes") or {}, sort_keys=True),
        "screen_result": vendor.get("screen_result"),
        "screen_reasons": json.dumps(vendor["screen_reasons"], sort_keys=True) if vendor.get("screen_reasons") is not None else None,
        "status": INITIAL_STATUS,
        "status_confidence": vendor.get("status_confidence"),
        "coverage_confidence": vendor.get("coverage_confidence"),
        "next_action": vendor.get("next_action"),
        "owner": vendor.get("owner"),
        "due_at": vendor.get("due_at"),
        "first_seen_run_id": vendor.get("first_seen_run_id"),
        "discovered_via": vendor.get("discovered_via"),
        "discovered_at": vendor.get("discovered_at") or now,
        "updated_at": now,
    }
    with conn:
        conn.execute(
            f"INSERT INTO vendors ({', '.join(cols)}) VALUES ({', '.join('?' for _ in cols)})",
            tuple(cols.values()),
        )
        conn.execute(
            "INSERT INTO events (vendor_id, from_status, to_status, from_stage, to_stage, actor, reason, "
            "confidence, evidence_ref, payload, created_at) VALUES (?, NULL, ?, NULL, NULL, ?, ?, NULL, NULL, ?, ?)",
            (vendor["vendor_id"], INITIAL_STATUS, actor, reason,
             json.dumps(payload, sort_keys=True) if payload else None, now),
        )
    return True


def transition(vendor_id: str, to_status: str, actor: str, reason: str,
               confidence: float | None = None, evidence_ref: str | int | None = None,
               to_stage: str | None = None, payload: dict[str, Any] | None = None,
               conn: sqlite3.Connection | None = None) -> int:
    """Move a vendor to `to_status` (and/or `to_stage`) by appending an events row.

    Raises IllegalTransition for moves outside PRD Section 7 and GateViolation
    when the actor may not perform the move.  Returns the new event_id.
    """
    conn = conn or get_conn()
    if to_status not in STATUSES:
        raise IllegalTransition(f"unknown status {to_status!r}")
    if to_stage is not None and to_stage not in STAGES:
        raise IllegalTransition(f"unknown diligence_stage {to_stage!r}")
    from_status, from_stage = current_state(vendor_id, conn)
    if to_status == "In Discussion" and to_stage is None and from_status != "In Discussion":
        to_stage = STAGES[0]  # entering In Discussion starts at 'responded'
    if not _is_legal(from_status, to_status, from_stage, to_stage):
        raise IllegalTransition(
            f"{vendor_id}: {from_status}/{from_stage} -> {to_status}/{to_stage} is not in the whitelist")
    _check_gate(from_status, to_status, to_stage, actor, confidence, reason)
    now = utcnow()
    with conn:
        cur = conn.execute(
            "INSERT INTO events (vendor_id, from_status, to_status, from_stage, to_stage, actor, reason, "
            "confidence, evidence_ref, payload, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            (vendor_id, from_status, to_status, from_stage, to_stage, actor, reason, confidence,
             str(evidence_ref) if evidence_ref is not None else None,
             json.dumps(payload, sort_keys=True) if payload else None, now),
        )
        conn.execute(
            "UPDATE vendors SET status = ?, status_confidence = ?, updated_at = ? WHERE vendor_id = ?",
            (to_status, confidence, now, vendor_id),
        )
    return int(cur.lastrowid)


def revert_event(event_id: int, actor: str = "human", reason: str = "reverted",
                 conn: sqlite3.Connection | None = None) -> int:
    """One-click revert of an auto-applied transition (PRD S4).

    Only the vendor's latest event can be reverted, only if it was not made by a
    human, and only by a human.  The revert is itself an events row (payload
    {"reverts": event_id}) so the log stays append-only and replayable.
    """
    conn = conn or get_conn()
    if actor != "human":
        raise GateViolation("reverts are human-only")
    if not reason or not str(reason).strip():
        raise GateViolation("a non-empty reason is required for a revert")
    ev = conn.execute("SELECT * FROM events WHERE event_id = ?", (event_id,)).fetchone()
    if ev is None:
        raise KeyError(f"unknown event_id {event_id}")
    latest = conn.execute(
        "SELECT event_id FROM events WHERE vendor_id = ? ORDER BY created_at DESC, event_id DESC LIMIT 1",
        (ev["vendor_id"],)).fetchone()
    if latest["event_id"] != event_id:
        raise IllegalTransition(f"event {event_id} is not the latest event for {ev['vendor_id']}; cannot revert")
    if ev["actor"] == "human":
        raise GateViolation("only auto-applied (non-human) transitions can be reverted")
    if ev["from_status"] is None:
        raise IllegalTransition("the creation event cannot be reverted")
    now = utcnow()
    with conn:
        cur = conn.execute(
            "INSERT INTO events (vendor_id, from_status, to_status, from_stage, to_stage, actor, reason, "
            "confidence, evidence_ref, payload, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, NULL, ?, ?, ?)",
            (ev["vendor_id"], ev["to_status"], ev["from_status"], ev["to_stage"], ev["from_stage"], actor, reason,
             ev["evidence_ref"], json.dumps({"reverts": event_id}), now),
        )
        conn.execute("UPDATE vendors SET status = ?, status_confidence = NULL, updated_at = ? WHERE vendor_id = ?",
                     (ev["from_status"], now, ev["vendor_id"]))
    return int(cur.lastrowid)


def _reverts(payload: str | None) -> int | None:
    if not payload:
        return None
    try:
        data = json.loads(payload)
    except ValueError:
        return None
    return data.get("reverts") if isinstance(data, dict) else None


def rebuild_status(vendor_id: str, conn: sqlite3.Connection | None = None) -> str:
    """Replay the event log for one vendor and assert it matches vendors.status."""
    conn = conn or get_conn()
    stored = conn.execute("SELECT status FROM vendors WHERE vendor_id = ?", (vendor_id,)).fetchone()
    if stored is None:
        raise KeyError(f"unknown vendor_id {vendor_id!r}")
    rows = conn.execute(
        "SELECT event_id, from_status, to_status, from_stage, to_stage, payload FROM events "
        "WHERE vendor_id = ? ORDER BY created_at, event_id", (vendor_id,)).fetchall()
    if not rows:
        raise StatusMismatch(f"{vendor_id}: no events, cannot reconstruct status")
    by_id = {ev["event_id"]: ev for ev in rows}
    status: str | None = None
    stage: str | None = None
    for ev in rows:
        if ev["from_status"] != status:
            raise StatusMismatch(
                f"{vendor_id}: event {ev['event_id']} starts from {ev['from_status']!r} but replay is at {status!r}")
        ref_id = _reverts(ev["payload"])
        if ref_id is not None:
            ref = by_id.get(ref_id)
            legal = (ref is not None and ref["to_status"] == ev["from_status"] and ref["from_status"] == ev["to_status"]
                     and ref["to_stage"] == ev["from_stage"] and ref["from_stage"] == ev["to_stage"])
        else:
            legal = _is_legal(status, ev["to_status"], stage, ev["to_stage"])
        if not legal:
            raise StatusMismatch(
                f"{vendor_id}: event {ev['event_id']} {status} -> {ev['to_status']} is illegal")
        status = ev["to_status"]
        stage = ev["to_stage"] if status == "In Discussion" else None
    if status != stored["status"]:
        raise StatusMismatch(f"{vendor_id}: replayed {status!r} != stored {stored['status']!r}")
    return status


def history(vendor_id: str, conn: sqlite3.Connection | None = None) -> list[sqlite3.Row]:
    conn = conn or get_conn()
    return conn.execute("SELECT * FROM events WHERE vendor_id = ? ORDER BY created_at, event_id", (vendor_id,)).fetchall()


def rebuild_all(conn: sqlite3.Connection | None = None) -> dict[str, str]:
    conn = conn or get_conn()
    ids = [r["vendor_id"] for r in conn.execute("SELECT vendor_id FROM vendors ORDER BY vendor_id")]
    return {vid: rebuild_status(vid, conn) for vid in ids}

File: src/lib/db/state.ts
/**
 * The only way status / diligence_stage may change. `transition()` validates
 * against the PRD §6 whitelist, appends an events row and updates vendors in
 * one transaction. `rebuildStatus()` replays events and throws on mismatch.
 */
import { asc, desc, eq } from "drizzle-orm";

import { getDb, nowIso, type DbOrTx } from "./index";
import {
  ACTORS,
  DILIGENCE_STAGES,
  VENDOR_STATUSES,
  events,
  vendors,
  type Actor,
  type DiligenceStage,
  type VendorStatus,
} from "./schema";
import type { EventRow } from "./schema";

export type TransitionErrorCode =
  | "not_found"
  | "invalid_input"
  | "illegal_transition"
  | "illegal_actor"
  | "illegal_stage"
  | "confidence_below_threshold"
  | "reason_required"
  | "no_change";

export class TransitionError extends Error {
  readonly code: TransitionErrorCode;
  constructor(code: TransitionErrorCode, message: string) {
    super(message);
    this.name = "TransitionError";
    this.code = code;
  }
}

export class StateMismatchError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "StateMismatchError";
  }
}

export const LLM_AUTO_APPLY_THRESHOLD = 0.85;
/** Stages that always go through a human, never llm_inference. */
export const HUMAN_ONLY_STAGES: readonly DiligenceStage[] = [
  "quote_received",
  "sampling",
];

export type TransitionInput = {
  vendorId: string;
  toStatus: VendorStatus;
  actor: Actor;
  reason: string;
  confidence?: number;
  evidenceRef?: string;
  /** Only meaningful when toStatus is "In Discussion". */
  toStage?: DiligenceStage | null;
  /** Free-form extra data stored on the events row. */
  payload?: Record<string, unknown>;
};

export type TransitionResult = {
  eventId: number;
  vendorId: string;
  fromStatus: VendorStatus;
  toStatus: VendorStatus;
  fromStage: DiligenceStage | null;
  toStage: DiligenceStage | null;
};

type Rule = {
  from: VendorStatus | "*";
  to: VendorStatus;
  actors: readonly Actor[];
  requiresReason?: boolean;
};

/** PRD §6. `*` = any status. */
export const TRANSITION_WHITELIST: readonly Rule[] = [
  { from: "Identified", to: "Screened", actors: ["system", "adapter"] },
  { from: "Screened", to: "Qualified", actors: ["human"] },
  { from: "Screened", to: "Rejected", actors: ["human"], requiresReason: true },
  { from: "Qualified", to: "Contacted", actors: ["human"] },
  { from: "Qualified", to: "Rejected", actors: ["human"], requiresReason: true },
  { from: "Contacted", to: "Replied", actors: ["system"] },
  { from: "Contacted", to: "Dormant", actors: ["system"] },
  { from: "Replied", to: "In Discussion", actors: ["llm_inference", "human"] },
  { from: "Replied", to: "Rejected", actors: ["llm_inference", "human"] },
  // stage change while In Discussion
  { from: "In Discussion", to: "In Discussion", actors: ["llm_inference", "human"] },
  { from: "In Discussion", to: "Rejected", actors: ["llm_inference", "human"] },
  { from: "*", to: "Approved", actors: ["human"] },
  { from: "Rejected", to: "Qualified", actors: ["human"], requiresReason: true },
  { from: "Dormant", to: "Qualified", actors: ["human"], requiresReason: true },
];

export const REVERT_REASON = "revert";

/** Pure whitelist check (no DB): may `actor` move a vendor from `from` to `to`? */
export function isLegalTransition(from: VendorStatus, to: VendorStatus, actor: Actor): boolean {
  const rule = findRule(from, to);
  return Boolean(rule && rule.actors.includes(actor));
}

function stageIndex(stage: DiligenceStage | null): number {
  return stage === null ? -1 : DILIGENCE_STAGES.indexOf(stage);
}

function findRule(
  from: VendorStatus,
  to: VendorStatus,
): Rule | undefined {
  return (
    TRANSITION_WHITELIST.find((r) => r.from === from && r.to === to) ??
    TRANSITION_WHITELIST.find((r) => r.from === "*" && r.to === to)
  );
}

function validateInput(input: TransitionInput): void {
  if (!input.vendorId) {
    throw new TransitionError("invalid_input", "vendorId is required");
  }
  if (!VENDOR_STATUSES.includes(input.toStatus)) {
    throw new TransitionError(
      "invalid_input",
      `unknown status "${String(input.toStatus)}"`,
    );
  }
  if (!ACTORS.includes(input.actor)) {
    throw new TransitionError(
      "invalid_input",
      `unknown actor "${String(input.actor)}"`,
    );
  }
  if (typeof input.reason !== "string" || input.reason.trim() === "") {
    throw new TransitionError("reason_required", "reason is required");
  }
  if (
    input.toStage !== undefined &&
    input.toStage !== null &&
    !DILIGENCE_STAGES.includes(input.toStage)
  ) {
    throw new TransitionError(
      "invalid_input",
      `unknown stage "${String(input.toStage)}"`,
    );
  }
  if (
    input.confidence !== undefined &&
    (Number.isNaN(input.confidence) ||
      input.confidence < 0 ||
      input.confidence > 1)
  ) {
    throw new TransitionError(
      "invalid_input",
      "confidence must be between 0 and 1",
    );
  }
}

/**
 * Apply one status/stage change. Throws TransitionError on anything the
 * PRD §6 whitelist does not allow. Pass a transaction handle to compose
 * with other writes.
 */
export function transition(
  input: TransitionInput,
  db: DbOrTx = getDb(),
): TransitionResult {
  validateInput(input);

  return db.transaction((tx) => {
    const vendor = tx
      .select({
        status: vendors.status,
        diligence_stage: vendors.diligence_stage,
      })
      .from(vendors)
      .where(eq(vendors.vendor_id, input.vendorId))
      .get();
    if (!vendor) {
      throw new TransitionError(
        "not_found",
        `vendor "${input.vendorId}" not found`,
      );
    }

    const fromStatus = vendor.status;
    const fromStage = vendor.diligence_stage ?? null;
    const { toStatus, actor } = input;

    const isRevert =
      actor === "human" && input.reason.trim().toLowerCase() === REVERT_REASON;
    const revertTarget = isRevert ? resolveRevertTarget(tx, input) : null;

    // Resolve the target stage.
    let toStage: DiligenceStage | null;
    if (isRevert) {
      if (!revertTarget) {
        throw new TransitionError(
          "illegal_transition",
          "revert requires a prior llm_inference event to reverse",
        );
      }
      // A revert restores exactly the state the reverted event started from.
      toStage = revertTarget.from_stage ?? null;
    } else if (input.toStage !== undefined && input.toStage !== null) {
      if (toStatus !== "In Discussion") {
        throw new TransitionError(
          "illegal_stage",
          `a diligence stage can only be set while "In Discussion" (got ${toStatus})`,
        );
      }
      toStage = input.toStage;
    } else if (toStatus === "In Discussion") {
      toStage = fromStatus === "In Discussion" ? fromStage : "responded";
    } else {
      // Leaving or outside In Discussion: the stage is carried unchanged.
      toStage = input.toStage === null ? null : fromStage;
    }

    if (isRevert && revertTarget) {
      // Reverse of a prior llm_inference event (PRD §4.3 "Revert").
      const target = revertTarget;
      if (
        target.to_status !== fromStatus ||
        (target.to_stage ?? null) !== fromStage
      ) {
        throw new TransitionError(
          "illegal_transition",
          `event ${target.event_id} led to ${target.to_status}/${target.to_stage ?? "-"} but the vendor is now at ${fromStatus}/${fromStage ?? "-"}`,
        );
      }
      if (target.from_status !== toStatus) {
        throw new TransitionError(
          "illegal_transition",
          `revert of event ${target.event_id} must return to ${target.from_status}/${target.from_stage ?? "-"}`,
        );
      }
    } else {
      const rule = findRule(fromStatus, toStatus);
      if (!rule) {
        throw new TransitionError(
          "illegal_transition",
          `${fromStatus} -> ${toStatus} is not allowed`,
        );
      }
      if (!rule.actors.includes(actor)) {
        throw new TransitionError(
          "illegal_actor",
          `${fromStatus} -> ${toStatus} may not be performed by ${actor} (allowed: ${rule.actors.join(", ")})`,
        );
      }
      if (fromStatus === toStatus && fromStage === toStage) {
        throw new TransitionError(
          "no_change",
          `vendor is already ${fromStatus}/${fromStage ?? "-"}`,
        );
      }
      if (actor === "llm_inference") {
        if (input.confidence === undefined) {
          throw new TransitionError(
            "invalid_input",
            "llm_inference transitions require a confidence",
          );
        }
        if (input.confidence < LLM_AUTO_APPLY_THRESHOLD) {
          throw new TransitionError(
            "confidence_below_threshold",
            `confidence ${input.confidence} is below the auto-apply threshold ${LLM_AUTO_APPLY_THRESHOLD}; file a proposal instead`,
          );
        }
        if (toStage !== null && HUMAN_ONLY_STAGES.includes(toStage)) {
          throw new TransitionError(
            "illegal_stage",
            `stage ${toStage} is always a human decision; file a proposal instead`,
          );
        }
        if (
          toStatus === "In Discussion" &&
          fromStatus === "In Discussion" &&
          stageIndex(toStage) <= stageIndex(fromStage)
        ) {
          throw new TransitionError(
            "illegal_stage",
            `llm_inference may only advance the stage (${fromStage ?? "-"} -> ${toStage ?? "-"})`,
          );
        }
      }
    }

    const createdAt = nowIso();
    const inserted = tx
      .insert(events)
      .values({
        vendor_id: input.vendorId,
        from_status: fromStatus,
        to_status: toStatus,
        from_stage: fromStage,
        to_stage: toStage,
        actor,
        reason: input.reason,
        confidence: input.confidence ?? null,
        evidence_ref: input.evidenceRef ?? null,
        payload: input.payload ?? null,
        created_at: createdAt,
      })
      .returning({ event_id: events.event_id })
      .get();

    tx.update(vendors)
      .set({
        status: toStatus,
        diligence_stage: toStage,
        status_confidence: input.confidence ?? null,
        updated_at: createdAt,
      })
      .where(eq(vendors.vendor_id, input.vendorId))
      .run();

    return {
      eventId: inserted.event_id,
      vendorId: input.vendorId,
      fromStatus,
      toStatus,
      fromStage,
      toStage,
    };
  });
}

function resolveRevertTarget(db: DbOrTx, input: TransitionInput) {
  const explicit = input.payload?.revert_event_id;
  if (typeof explicit === "number") {
    const row = db
      .select()
      .from(events)
      .where(eq(events.event_id, explicit))
      .get();
    if (!row || row.vendor_id !== input.vendorId || row.actor !== "llm_inference") {
      return null;
    }
    return row;
  }
  const latest = db
    .select()
    .from(events)
    .where(eq(events.vendor_id, input.vendorId))
    .orderBy(desc(events.event_id))
    .limit(1)
    .get();
  return latest && latest.actor === "llm_inference" ? latest : null;
}

/**
 * Informational event (tag_changed, screen_changed): appended to the log
 * without changing status or stage, so replay treats it as identity.
 */
export function recordEvent(
  input: { vendorId: string; actor: Actor; reason: string; payload?: Record<string, unknown>; evidenceRef?: string },
  db: DbOrTx = getDb(),
): EventRow {
  const v = db.select().from(vendors).where(eq(vendors.vendor_id, input.vendorId)).get();
  if (!v) throw new TransitionError("not_found", `vendor ${input.vendorId} not found`);
  return db
    .insert(events)
    .values({
      vendor_id: input.vendorId,
      from_status: v.status,
      to_status: v.status,
      from_stage: v.diligence_stage,
      to_stage: v.diligence_stage,
      actor: input.actor,
      reason: input.reason,
      confidence: null,
      evidence_ref: input.evidenceRef ?? null,
      payload: input.payload ?? null,
      created_at: nowIso(),
    })
    .returning()
    .get();
}

export type RebuildResult = {
  vendorId: string;
  status: VendorStatus;
  stage: DiligenceStage | null;
  eventCount: number;
};

/**
 * Replay the append-only event log from the initial state (Identified, no
 * stage). Throws StateMismatchError if the chain is broken or the replayed
 * state differs from vendors.status / diligence_stage.
 */
export function rebuildStatus(vendorId: string, db: DbOrTx = getDb()): RebuildResult {
  const vendor = db
    .select({
      status: vendors.status,
      diligence_stage: vendors.diligence_stage,
    })
    .from(vendors)
    .where(eq(vendors.vendor_id, vendorId))
    .get();
  if (!vendor) {
    throw new TransitionError("not_found", `vendor "${vendorId}" not found`);
  }

  const rows = db
    .select()
    .from(events)
    .where(eq(events.vendor_id, vendorId))
    .orderBy(asc(events.event_id))
    .all();

  let status: VendorStatus = "Identified";
  let stage: DiligenceStage | null = null;
  for (const e of rows) {
    if (e.from_status !== status || (e.from_stage ?? null) !== stage) {
      throw new StateMismatchError(
        `vendor ${vendorId}: event ${e.event_id} starts from ${e.from_status}/${e.from_stage ?? "-"} but replay is at ${status}/${stage ?? "-"}`,
      );
    }
    status = e.to_status;
    stage = e.to_stage ?? null;
  }

  if (vendor.status !== status || (vendor.diligence_stage ?? null) !== stage) {
    throw new StateMismatchError(
      `vendor ${vendorId}: vendors row says ${vendor.status}/${vendor.diligence_stage ?? "-"} but events replay to ${status}/${stage ?? "-"}`,
    );
  }

  return { vendorId, status, stage, eventCount: rows.length };
}
